### PR TITLE
Enhancement: Improve Security Against Malicious Delegatecall

### PR DESCRIPTION
**Pull request by asiniscalchi**
As described here: https://medium.com/@Beosin_com/security-audit-series-what-is-a-precompiled-contract-vulnerability-5174e20e24e9

### CODE REVIEW

**Review by ccubu — APPROVED** [review]

File: ownership-chain/runtime/src/precompiles/mod.rs [hunk 1]
```diff
@@ -26,6 +27,27 @@ where
...
 	}
+
+	fn is_delegatecall_to_custom_precompile(
+		&self,
+		code_address: H160,
+		context_address: H160,
+	) -> bool {
+		// Check if the code address is a precompile
+		if let IsPrecompileResult::Answer { is_precompile, .. } =
+			self.is_precompile(code_address, u64::MAX)
```
Comment (magecnion) [inline ownership-chain/runtime/src/precompiles/mod.rs:38]: why `u64::MAX` ? could it be removed as we don't use it within `is_precompile` fn?

File: ownership-chain/runtime/Cargo.toml [hunk 2]
```diff
@@ -90,6 +90,8 @@ pallet-evm-precompile-simple = { workspace = true }
 
 [dev-dependencies]
 static_assertions = { workspace = true }
+precompile-utils = { workspace = true, features = ["testing"] }
+evm = { workspace = true }
```
Comment (magecnion) [inline ownership-chain/runtime/Cargo.toml:94]: what about include also `evm` dependency when testing as `precompile-utils` ?

**Review by magecnion — CHANGES_REQUESTED** [review]

File: ownership-chain/runtime/Cargo.toml (hunk 2 above)
Comment (magecnion) [reply ownership-chain/runtime/Cargo.toml:94]: disregard, conceptual misunderstanding by my side

File: ownership-chain/runtime/src/precompiles/mod.rs (hunk 1 above)
Comment (asiniscalchi) [reply ownership-chain/runtime/src/precompiles/mod.rs:38]: it's not possible to remove the `_gas` parameter of the function `is_precompile` cause it is declared in trait `PrecompileSet`

File: ownership-chain/runtime/src/precompiles/test.rs [hunk 3]
```diff
@@ -24,8 +23,84 @@ fn ethereum_precompiled_addresses_are_precompile() {
...
+	let context_address = hash(12);
+
+	// Verify that the code address is not a precompile and that it's not treated as a custom
+	// precompile delegate call
+	assert!(!is_precompile(code_address).unwrap());
+	assert!(!precompiles.is_delegatecall_to_custom_precompile(code_address, context_address));
+}
+
+#[test]
+/// Test to ensure that delegate calls to non-standard Ethereum precompile addresses are recognized.
```
Comment (dastanfv) [inline ownership-chain/runtime/src/precompiles/test.rs:50]: same here

File: ownership-chain/runtime/src/precompiles/test.rs [hunk 4]
```diff
@@ -24,8 +23,84 @@ fn ethereum_precompiled_addresses_are_precompile() {
...
+	let context_address = hash(123456);
+
+	// Verify that the code address is a precompile and is recognized as a custom precompile
+	// delegate call
+	assert!(is_precompile(code_address).unwrap());
+	assert!(precompiles.is_delegatecall_to_custom_precompile(code_address, context_address));
+}
+
+#[test]
+/// Test to ensure that delegate calls to standard Ethereum precompile addresses are not recognized
```
Comment (dastanfv) [inline ownership-chain/runtime/src/precompiles/test.rs:65]: same here

**Review by dastanfv — COMMENTED** [review]
very small comments

File: ownership-chain/runtime/Cargo.toml (hunk 2 above)
Comment (asiniscalchi) [reply ownership-chain/runtime/Cargo.toml:94]: `evm` and `precompile-utils` creates are already included only in tests (aka `[dev-dependecies]).

**Review by magecnion — APPROVED** [review]

File: ownership-chain/runtime/src/precompiles/test.rs [hunk 5]
```diff
@@ -24,8 +23,84 @@ fn ethereum_precompiled_addresses_are_precompile() {
...
 	assert!(is_precompile(hash(5)).unwrap());
+	assert!(!is_precompile(hash(6)).unwrap());
 	assert!(!is_precompile(hash(1026)).unwrap());
 	assert!(is_precompile(hash(1027)).unwrap());
 	assert!(is_precompile(H160::from_str("0xfffffffffffffffffffffffe0000000000000005").unwrap())
 		.unwrap());
 }
+
+#[test]
+/// Test to ensure that delegate calls to addresses that are not precompiles are recognized
```
Comment (dastanfv) [inline ownership-chain/runtime/src/precompiles/test.rs:34]: comment above macro attribute looks cleaner and this is what we do in the rest of the codebase. So would suggest to re-order.

Comment (asiniscalchi) [reply ownership-chain/runtime/src/precompiles/test.rs:34]: sgtm

**Review by dastanfv — APPROVED** [review]

**Pull request merged**